nat-lab/tests/test_core_api.py

```diff
@@ -0,0 +1,413 @@
+                value, expected_type
+            ), f"Key '{key}' should be of type {expected_type}, but got {type(value)}"
+
+        else:
+            raise ValueError(f"Unexpected type for key '{key}': {expected_type}")
+
+
+def verify_uuid(uuid_to_test, version=4):
+    try:
+        uuid_obj = UUID(uuid_to_test, version=version)
+    except ValueError:
+        assert False, "Not a valid UUID"
+    assert str(uuid_obj) == uuid_to_test, "Not a valid UUID"
+
+
+def create_machine_data(public_key, hardware_identifier, os, os_version):
+    return MachineData(
+        public_key=public_key,
+        hardware_identifier=hardware_identifier,
+        os=os,
+        os_version=os_version,
+    )
+
+
+async def send_http_request(
+    connection, endpoint, method, data=None, expect_response=True
+):
+    curl_command = [
+        "curl",
+        "-X",
+        method,
+        endpoint,
+        "-H",
+        "Content-Type: application/json",
+    ]
+
+    if data:
+        curl_command.extend(["-d", data])
+
+    process = await connection.create_process(curl_command).execute()
+    response = process.get_stdout()
+    if expect_response:
+        try:
+            return json.loads(response)
+        except json.JSONDecodeError:
+            assert False, f"Expected JSON response but got: {response}"
+    return None
+
+
+async def clean_up_machines(connection, server_host):
+    machines = await send_http_request(
+        connection, f"{server_host}/v1/meshnet/machines", method="GET"
+    )
+
+    for machine in machines:
+        await send_http_request(
+            connection,
+            f"{server_host}/v1/meshnet/machines/{machine['identifier']}",
+            method="DELETE",
+            expect_response=False,
+        )
+
+
+@pytest.fixture(name="machine_data")
+def fixture_machine_data(request):
+    return request.param
+
+
+@pytest.fixture(name="server_host")
+def fixture_server_host():
+    return "http://10.0.80.86:8080"
+
+
+@pytest.fixture(name="registered_machines")
+async def fixture_register_machine(server_host, machine_data):
+    async with AsyncExitStack() as exit_stack:
+        connection = await exit_stack.enter_async_context(
+            new_connection_by_tag(ConnectionTag.DOCKER_CONE_CLIENT_1)
+        )
+
+        await clean_up_machines(connection, server_host)
+
+        registered_machines = []
+
+        for data in machine_data:
+            payload = json.dumps(data.__dict__)
+
+            response_data = await send_http_request(
+                connection,
+                f"{server_host}/v1/meshnet/machines",
+                method="POST",
+                data=payload,
+            )
+            registered_machines.append(response_data)
+
+        return registered_machines
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "machine_data",
+    [[
+        create_machine_data(
+            generate_wireguard_like_public_key(),
+            generate_hardware_identifier(),
```

Comment:
Nit: I'm not sure about this, but maybe it would be better to fix sth here? I guess test runs should be as similar as possible.